### PR TITLE
Rate Limit Configuration: Header value match action

### PR DESCRIPTION
**Pull request by ccaraman**
- Move ConfigUtility into its own config_utility.h/cc.
- Add Header Value Match Action that allows for a rate limit descriptor to be sent if a request matches a set of  header values specified in the config. 
- Create base class Json::JsonValidator for classes to inherit from when they want to validate a config schema prior to initializing member variables.

### CODE REVIEW

**Comment by ccaraman** [comment]
@lyft/network-team

File: docs/configuration/http_conn_man/route_config/rate_limits.rst [hunk 1]
```diff
@@ -141,6 +141,57 @@ The following descriptor entry is appended to the descriptor:
...
+:ref:`headers<config_http_conn_man_route_table_rate_limit_headers>`
+    *(required, array)* Specifies a set of headers that the rate limit action should match on.
+
+The following descriptor entry is appended to the descriptor if the request matches the headers
+specified in the action config:
+
+    * ("header_match", "<descriptor_value>")
+
+
+.. _config_http_conn_man_route_table_rate_limit_headers:
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/rate_limits.rst]: I think we now have this text in potentially 3 places in the docs. Can we de-dup this and just link to it from all 3 places?

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -611,6 +611,32 @@ const std::string Json::Schema::HTTP_RATE_LIMITS_CONFIGURATION_SCHEMA(R"EOF(
...
+      },
+      "header_value_match" : {
+        "type" : "object",
+        "properties" : {
+          "type" : {
+            "type" : "string",
+            "enum" : ["header_value_match"]
+          },
+          "descriptor_value" : {"type" : "string"},
+          "headers" : {
```
Comment (mattklein123) [inline source/common/json/config_schemas.cc]: This is also duplicated several times now. Can we de-dup and do a sub-schema check in the ConfigUtility code?

File: source/common/router/config_utility.cc [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#include "config_utility.h"
```
Comment (mattklein123) [inline source/common/router/config_utility.cc]: Did you change any code in config_utility or is it just copy/paste to a new file without any changes?

Comment (ccaraman) [inline source/common/router/config_utility.cc]: Copy Paste

Comment (mattklein123) [inline source/common/router/config_utility.cc]: OK not a big deal but in the future I would do a separate code move PR, just easier to review.

File: source/common/json/json_validator.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "envoy/json/json_object.h"
+
+namespace Json {
+
+class JsonValidator {
+public:
+  /*
```
Comment (mattklein123) [inline source/common/json/json_validator.h]: nit: move this comment to class level, you are also missing a '*' in the comment opening.

File: source/common/router/config_utility.h [hunk 5]
```diff
@@ -12,10 +15,15 @@ namespace Router {
  */
 class ConfigUtility {
 public:
-  struct HeaderData {
-    HeaderData(const Http::LowerCaseString& name, const std::string& value, const bool is_regex)
-        : name_(name), value_(value), regex_pattern_(value_, std::regex::optimize),
-          is_regex_(is_regex) {}
+  struct HeaderData : public Json::JsonValidator {
```
Comment (mattklein123) [inline source/common/router/config_utility.h]: nit: public inheritance not needed

File: docs/configuration/http_conn_man/route_config/rate_limits.rst [hunk 6]
```diff
@@ -141,6 +141,30 @@ The following descriptor entry is appended to the descriptor:
...
+  {
+    "type": "header_value_match",
+    "descriptor_value" : "...",
+    "headers" : []
+  }
+
+
+descriptor_value
+    *(required, string)* The value to use in the descriptor entry.
+:ref:`headers<config_http_conn_man_route_table_route_headers>`
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/rate_limits.rst]: nit: newline before this line

**Pull request merged**